added/improved provider documentation

redhatHameed opened the pull request:
## Description
<!-- Please include a summary of the change and link to the related Jira issue. Please add any additional motivation and context as needed. Screenshots are also welcome -->

## Verification Steps
<!--
Add the steps required to check this change. Following an example.

1. Go to `XX >> YY >> SS`
2. Create a new item `N` with the info `X`
3. Try to edit this item 
4. Check if in the left menu the feature X is not so long present.
-->

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist (Definition of Done)
<!-- Please strikethrough options not relevant using two tildes ~~Text~~. Do not delete non relevant options -->
- [ ] All acceptance criteria specified in JIRA or in issue number have been completed
- [ ] Unit tests added that prove the fix is effective or the feature works 
- [ ] Documentation added for the feature
- [ ] all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,557 @@
+
+
+## Goals:
+- Allow Database Partner Provider operators to register themselves with the Database-as-a-Service operator via DBaaSProvider resource (CR)
```
Database-as-a-Service => Database-as-a-Service (DBaaS)
DBaaSProvider resource => DBaaSProvider custom resource

redhatHameed replied:
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,557 @@
+
+## Goals:
+- Allow Database Partner Provider operators to register themselves with the Database-as-a-Service operator via DBaaSProvider resource (CR)
+- Query a provider’s system via their own operator to determine a list of all databases available for the given user
```
"Query a provider’s system via their own operator"=>"Query the provider's cloud-hosted database services via API calls"

redhatHameed replied:
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,557 @@
+## Goals:
+- Allow Database Partner Provider operators to register themselves with the Database-as-a-Service operator via DBaaSProvider resource (CR)
+- Query a provider’s system via their own operator to determine a list of all databases available for the given user
+- Allow OpenShift developers to retrieve all information from the provider’s inventory response required for connecting to database instances (connection string, credentials, etc.)
```
"Allow OpenShift developers to retrieve all information from the provider’s inventory response"=>
"Allow applications to retrieve all information from the provider's database cloud-hosting system"

redhatHameed replied:
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,557 @@
+- Allow Database Partner Provider operators to register themselves with the Database-as-a-Service operator via DBaaSProvider resource (CR)
+- Query a provider’s system via their own operator to determine a list of all databases available for the given user
+- Allow OpenShift developers to retrieve all information from the provider’s inventory response required for connecting to database instances (connection string, credentials, etc.)
+- Allows Service Admin/Developer provisioning of trial or small database clusters by provider operator
```
"Allows Service Admin/Developer provisioning of trial or small database clusters" => Allows Service Admin/Developers to provision trial or dedicated database clusters (or instances)

redhatHameed replied:
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,557 @@
+- Allow OpenShift developers to retrieve all information from the provider’s inventory response required for connecting to database instances (connection string, credentials, etc.)
+- Allows Service Admin/Developer provisioning of trial or small database clusters by provider operator
+
+## Non-Goals:
```
This is not needed from the user guide point of view

redhatHameed commented on docs/images/installed-openshift.png:
@jianrongzhang89 I moved this document from drive and I don't have actual diagram to change this at the moment. This is minor so let's keep the same.

redhatHameed commented on docs/images/inventory-request.png:
Let's keep the same. as actual diagram not present in the drive.

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,555 @@
+- Allow applications to retrieve all information from the provider's database cloud-hosting system required for connecting to database instances (connection string, credentials, etc.)
+- Allows Service Admin/Developers to provision trial or dedicated database clusters (or instances) by provider operator
+
+## References:
```
Move References to the end of this file.

redhatHameed replied:
Done

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,555 @@
+- All code blocks & samples herein are part of our [DBaaSProvider API](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/api/v1beta1/dbaasprovider.go) which details the struct types used for communicating & collaborating with Partner Providers.
+- The provider operator example :  https://github.com/RHEcosystemAppEng/provider-operator-example
+-
+## Phase Requisites:
```
Remove "Phase".

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,555 @@
+
+
+
+## Registering the Provider Operator:
```
Registering the Provider Operator =>Register the Provider Operator with DBaaS Operator

redhatHameed replied:
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,555 @@
+
+```
+
+- The DBaasProvider CR for example includes:
````
No need to describe the fields here. Instead, add a reference.

For more information about each field defined in the DBaaSProvider CR, see the documentation (https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/docs/api/markdown/ref.md#dbaasprovider)

redhatHameed replied:
Let's keep both added link for more details.

jianrongzhang89 replied:
OK

redhatHameed commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,555 @@
+
+## Registering the Provider Operator:
+
+- As a first step within OpenShift, the provider operator should place a **cluster-scoped** DBaaSProvider CR on the OpenShift cluster which will inform the DBaaS operator as to how to work with their operator without having any prior knowledge.
```
updated

redhatHameed commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,555 @@
+      displayName: ''
+  groupVersion: dbaas.redhat.com/v1beta1
+```
+mongodb-atlas-registration
````
updated

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,479 @@
+
+Once the DBaaS operator has reconciled a provider’s DBaaSProvider, collaboration can now occur between the operators via creating/updating the specified
+*inventoryKind*  resource type. The first area of coordination between the operators occurs when listing all of a user’s available database instances.
+In order to communicate the list request to the provider operator, the DBaaS Operator will create a resource of type *inventoryKind* for the provider operator to reconcile. The *inventoryKind* resource will have a single spec field, *CredentialsRef*, which points to an on-cluster Secret resource containing all the user credentials fields required by the provider operator to query their platform ( ...
```
"In order to communicate the list request to the provider operator," => "The actual instances discovery is done via the corresponding provider operator, and to that end, "

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,479 @@
+}
+
+```
+## Inventory Listing Response:
````
=>Instance Discovery Responses

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,479 @@
+	ServiceInfo map[string]string `json:"serviceInfo,omitempty"`
+}
+```
+- Once the inventory list has been provided, partner providers may choose to monitor and update as needed to reflect the status of their platform, though this is not required for MVP phase work.
````
inventory list => instance list

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,479 @@
+	ServiceInfo map[string]string `json:"serviceInfo,omitempty"`
+}
+```
+- Once the inventory list has been provided, partner providers may choose to monitor and update as needed to reflect the status of their platform, though this is not required for MVP phase work.
````
What do you mean here: "though this is not required for MVP phase work"?

redhatHameed replied:
removed it.

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,479 @@
+```
+- Once the inventory list has been provided, partner providers may choose to monitor and update as needed to reflect the status of their platform, though this is not required for MVP phase work.
+- **Note**: The provider operator’s response should list all database instances that are available for the provider administrator’s credentials
+  - If the provider’s entity hierarchy uses multiple organizations, we anticipate that the provider operator will restructure the status response presented so that the organization is additional information
````
can you give an example here?

redhatHameed replied:
added example including instance, connection, inventory

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,479 @@
+
+- **Note**: There is a limit on how many records can be returned (by default etcd limits the maximum data entry size to 1.5MB).  In the future, another Condition can be added to indicate that not all of the requested records can be returned.  In that case, the admin may provide credentials with a narrower scope, and/or we can add support for provider-specific filters
+
+## Instance Connectivity:
```
=>Connect to a Database Instance in Your Application

ritz303 commented on docs/provider-guide/dbaas-provider-guide.md:
````diff
@@ -0,0 +1,479 @@
+```
+
+
+Upon reconciliation, the provider operator should use the provided *InventoryRef* to identify what instance has been requested and provide any further information required for connectivity via the *connectionKind* resource’s status:
````
via --> using

ritz303 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,479 @@
+At this point, the DBaaS Operator has collated everything it needs to provide OpenShift developer users the information needed to connect to any of their imported instances. While the DBaaS operator, Service Binding Operator and OCP environment perform further work to present instance connectivity to the user in a simple-to-consume fashion, the partner provider is not required to take any further ...
+
+## Instance Provisioning:
+The DBaaS Operator also allows administrator & developer users to request instance provisioning within their respective UX workflows. The DBaaS operator will await user creation of a *DBaaSInstance* (via UX workflow) that indicates a database instance to be created in an inventory. After this is received, the DBaaS Operator will create an *instanceKind* resource (as defined in the provider’s cust ...
```
Same as my comment on line 334.

redhatHameed replied:
done

ritz303 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,479 @@
+
+
+
+The provider operator should use the provided *InventoryRef* to get the credential details for the inventory (i.e. from the *secret*) and create the instance that has been requested, and then update the details of cluster request via the *DBaaSInstanceStatus*. Information required within the status includes as seen below:
```
We shouldn't Latin abbreviations in technical documentation.

... (i.e. from the *secret*) ... --> ... the inventory, that is from the *secret*, and ...

ritz303 reviewed (changes requested):
@redhatHameed : Thanks a lot for writing up a draft of this doc!  I left you some feedback, please don't hesitate to yell at me if you have questions or need any clarifications.

redhatHameed commented:
@jianrongzhang89 @ritz303 Thanks for reviewing this, I addressed most of the comments, Please take another looks let me know if any further changes required.

tchughesiv commented:
we can merge as soon as @jianrongzhang89 approves

jianrongzhang89 commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,576 @@
+- Allow database partner provider operators to register themselves with the Database-as-a-Service(DBaaS) operator using DBaaSProvider custom resource (CR).
+- Query the provider's cloud-hosted database services using API calls to determine a list of all databases available for the given user.
+- Allow applications to retrieve all information from the provider's database cloud-hosting system required for connecting to database instances such as connection string credentials.
+- Allows a Service Administrator or a Developer to provision a trial or dedicated database clusters or instances by provider operator.
```
provision a trial => provision trial

redhatHameed commented on docs/provider-guide/dbaas-provider-guide.md:
```diff
@@ -0,0 +1,576 @@
+## Goals:
+- Allow database partner provider operators to register themselves with the Database-as-a-Service(DBaaS) operator using DBaaSProvider custom resource (CR).
+- Query the provider's cloud-hosted database services using API calls to determine a list of all databases available for the given user.
+- Allow applications to retrieve all information from the provider's database cloud-hosting system required for connecting to database instances such as connection string credentials.
```
done

jianrongzhang89 reviewed (approved):
Looks good to me.
